### PR TITLE
GEODE-4685: Moving of static DefaultQuery.setPdxReadSerialized to the TypeRegistry.

### PR DESCRIPTION
**Pull request by kohlmu-pivotal**
GEODE-4685: Moving of static DefaultQuery.setPdxReadSerialized to the TypeRegistry.
Cleaned up the overriding of readSerialized to reset to previous value.
Added cache to AbstractJdbcCallback.java so that children classes can access it.
Replaced AtomicLong with LongAdder.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/cache/GemFireCacheImpl.java [hunk 1]
```diff
@@ -5151,7 +5156,7 @@ private void basicSetPdxSerializer(PdxSerializer serializer) {
   @Override
   public void setReadSerializedForCurrentThread(boolean value) {
     PdxInstanceImpl.setPdxReadSerialized(value);
-    DefaultQuery.setPdxReadSerialized(value);
+    this.getPdxRegistry().setPdxReadSerializedOverride(value);
```
Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/internal/cache/GemFireCacheImpl.java]: In the getPdxReadSerializedByAnyGemFireServices we check to see if pdx registry is null, but here we do not have an equivalent check.  Is there a way that the registry is null when setting?

File: geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/QueryMessage.java [hunk 2]
```diff
@@ -169,7 +169,9 @@ protected boolean operateOnPartitionedRegion(ClusterDistributionManager dm, Part
 
     DefaultQuery query = new DefaultQuery(this.queryString, pr.getCache(), false);
     // Remote query, use the PDX types in serialized form.
-    DefaultQuery.setPdxReadSerialized(pr.getCache(), true);
+    Boolean initialPdxReadSerialized =
```
Comment (jhuynh1) [inline geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/QueryMessage.java]: Same question, any chance the registry is null for these calls?

File: geode-core/src/main/java/org/apache/geode/cache/query/internal/DefaultQuery.java [hunk 3]
```diff
@@ -128,110 +126,106 @@
...
+      add("by");
+      add("as");
+    }
+  };
 
-  /**
-   * Caches the fields not found in any Pdx version. This threadlocal will be cleaned up after query
-   * execution completes in {@linkplain #executeUsingContext(ExecutionContext)}
-   */
+  // /**
```
Comment (upthewaterspout) [inline geode-core/src/main/java/org/apache/geode/cache/query/internal/DefaultQuery.java]: Commented out comment?

File: geode-core/src/main/java/org/apache/geode/distributed/internal/streaming/StreamingOperation.java [hunk 4]
```diff
@@ -209,12 +221,12 @@ public void process(DistributionMessage msg) {
         }
         if (isLast) {
           super.process(msg, false); // removes from members and cause us to
-                                     // ignore future messages received from that member
+          // ignore future messages received from that member
```
Comment (upthewaterspout) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/streaming/StreamingOperation.java]: Maybe take it easy with the auto format because it's generating a lot of unrelated whitespace changes.

File: geode-core/src/main/java/org/apache/geode/distributed/internal/streaming/StreamingOperation.java [hunk 5]
```diff
@@ -501,15 +517,23 @@ public void fromData(DataInput in) throws IOException, ClassNotFoundException {
       this.pdxReadSerialized = in.readBoolean();
       Version senderVersion = InternalDataSerializer.getVersionForDataStream(in);
       boolean isSenderAbove_8_1 = senderVersion.compareTo(Version.GFE_81) > 0;
+      InternalCache cache = null;
+      Boolean initialPdxReadSerialized = false;
+      try {
+        cache = (InternalCache) CacheFactory.getAnyInstance();
```
Comment (upthewaterspout) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/streaming/StreamingOperation.java]: Use GemFireCacheImpl.getForPdx. For two reasons - CacheFactory.getAnyInstance gets locks, so this could cause a deadlock. And getForPdx works while the cache is closing.

File: geode-core/src/main/java/org/apache/geode/pdx/internal/TypeRegistry.java [hunk 6]
```diff
@@ -65,6 +65,9 @@
 
   private final InternalCache cache;
 
+  private static final ThreadLocal<Boolean> pdxReadSerializedOverride =
```
Comment (upthewaterspout) [inline geode-core/src/main/java/org/apache/geode/pdx/internal/TypeRegistry.java]: It's a bit weird to go to all this trouble to use an instance of TypeRegistry with an instance method, only to end up updating a *static* threadlocal. Should this still be static?

Comment (kohlmu-pivotal) [reply geode-core/src/main/java/org/apache/geode/pdx/internal/TypeRegistry.java]: Thank you.. I overlooked that. Yes, the static is not required anymore.

File: geode-core/src/main/java/org/apache/geode/distributed/internal/streaming/StreamingOperation.java (hunk 5 above)
Comment (kohlmu-pivotal) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/streaming/StreamingOperation.java]: done

File: geode-core/src/main/java/org/apache/geode/distributed/internal/streaming/StreamingOperation.java (hunk 4 above)
Comment (kohlmu-pivotal) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/streaming/StreamingOperation.java]: I'll try and revert all whitespace changes

File: geode-core/src/main/java/org/apache/geode/cache/query/internal/DefaultQuery.java (hunk 3 above)
Comment (kohlmu-pivotal) [reply geode-core/src/main/java/org/apache/geode/cache/query/internal/DefaultQuery.java]: +1

File: geode-core/src/main/java/org/apache/geode/internal/cache/GemFireCacheImpl.java (hunk 1 above)
Comment (kohlmu-pivotal) [reply geode-core/src/main/java/org/apache/geode/internal/cache/GemFireCacheImpl.java]: there seems to be a possibility that the registry can be null. Will add the check.

File: geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/QueryMessage.java (hunk 2 above)
Comment (kohlmu-pivotal) [reply geode-core/src/main/java/org/apache/geode/internal/cache/partitioned/QueryMessage.java]: there seems to be a possibility that the registry can be null. Will add the check.

File: geode-core/src/main/java/org/apache/geode/distributed/internal/streaming/StreamingOperation.java (hunk 4 above)
Comment (upthewaterspout) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/streaming/StreamingOperation.java]: It's fine for this time, this was more of a suggestion for future diffs.

File: geode-connectors/src/test/java/org/apache/geode/connectors/jdbc/JdbcWriterIntegrationTest.java [hunk 7]
```diff
@@ -42,6 +42,7 @@
 import org.apache.geode.connectors.jdbc.internal.SqlHandler;
 import org.apache.geode.connectors.jdbc.internal.TestConfigService;
 import org.apache.geode.connectors.jdbc.internal.TestableConnectionManager;
+import org.apache.geode.internal.cache.InternalCache;
```
Comment (nabarunnag) [inline geode-connectors/src/test/java/org/apache/geode/connectors/jdbc/JdbcWriterIntegrationTest.java]: we can remove Cache import statement

File: geode-core/src/main/java/org/apache/geode/cache/query/internal/DefaultQuery.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (kohlmu-pivotal) [reply geode-core/src/main/java/org/apache/geode/cache/query/internal/DefaultQuery.java]: Commented out comments and the formatting applied on top.. never pretty... Already reverted and addressed

File: geode-lucene/src/test/java/org/apache/geode/cache/lucene/internal/LuceneEventListenerJUnitTest.java [hunk 8]
```diff
@@ -14,8 +14,21 @@
  */
 package org.apache.geode.cache.lucene.internal;
 
-import static org.junit.Assert.*;
-import static org.mockito.Mockito.*;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.fail;
```
Comment (nabarunnag) [inline geode-lucene/src/test/java/org/apache/geode/cache/lucene/internal/LuceneEventListenerJUnitTest.java]: Unused imports
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.spy;
import org.mockito.stubbing.Answer;
import org.apache.geode.cache.query.internal.DefaultQuery;

File: geode-core/src/main/java/org/apache/geode/distributed/internal/streaming/StreamingOperation.java [hunk 9]
```diff
@@ -501,15 +512,24 @@ public void fromData(DataInput in) throws IOException, ClassNotFoundException {
...
       Version senderVersion = InternalDataSerializer.getVersionForDataStream(in);
       boolean isSenderAbove_8_1 = senderVersion.compareTo(Version.GFE_81) > 0;
+      InternalCache cache = null;
+      Boolean initialPdxReadSerialized = false;
+      try {
+        cache =
+            (InternalCache) GemFireCacheImpl.getForPdx("fromData invocation in StreamingOperation");
+        initialPdxReadSerialized = cache.getPdxReadSerializedOverride();
+      } catch (CacheClosedException e) {
+
```
Comment (nabarunnag) [inline geode-core/src/main/java/org/apache/geode/distributed/internal/streaming/StreamingOperation.java]: Maybe we should add a log here. since we are consuming the exception silently.
Also i was wondering if we could just return from here. If the cache itself is closed there maybe no point in executing the below code. maybe.

File: geode-connectors/src/test/java/org/apache/geode/connectors/jdbc/JdbcLoaderIntegrationTest.java [hunk 10]
```diff
@@ -35,6 +35,7 @@
 import org.apache.geode.connectors.jdbc.internal.SqlHandler;
 import org.apache.geode.connectors.jdbc.internal.TestConfigService;
 import org.apache.geode.connectors.jdbc.internal.TestableConnectionManager;
+import org.apache.geode.internal.cache.InternalCache;
```
[the remark this answers is not in the document]
Comment (kohlmu-pivotal) [reply geode-connectors/src/test/java/org/apache/geode/connectors/jdbc/JdbcLoaderIntegrationTest.java]: +1

File: geode-connectors/src/test/java/org/apache/geode/connectors/jdbc/JdbcWriterIntegrationTest.java (hunk 7 above)
Comment (kohlmu-pivotal) [reply geode-connectors/src/test/java/org/apache/geode/connectors/jdbc/JdbcWriterIntegrationTest.java]: +1

File: geode-connectors/src/test/java/org/apache/geode/connectors/jdbc/JdbcWriterTest.java [hunk 11]
```diff
@@ -27,12 +27,15 @@
 import org.junit.Test;
 import org.junit.experimental.categories.Category;
 
+import org.apache.geode.cache.Cache;
```
[the remark this answers is not in the document]
Comment (kohlmu-pivotal) [reply geode-connectors/src/test/java/org/apache/geode/connectors/jdbc/JdbcWriterTest.java]: +1

File: geode-lucene/src/main/java/org/apache/geode/cache/lucene/internal/LuceneEventListener.java [hunk 12]
```diff
@@ -101,18 +107,18 @@ protected boolean process(final List<AsyncEvent> events) {
       }
       return true;
     } catch (BucketNotFoundException | RegionDestroyedException | PrimaryBucketException e) {
-      logger.debug("Bucket not found while saving to lucene index: " + e.getMessage(), e);
+      getLogger().debug("Bucket not found while saving to lucene index: " + e.getMessage(), e);
```
[the remark this answers is not in the document]
Comment (kohlmu-pivotal) [reply geode-lucene/src/main/java/org/apache/geode/cache/lucene/internal/LuceneEventListener.java]: done

File: geode-core/src/main/java/org/apache/geode/distributed/internal/streaming/StreamingOperation.java (hunk 9 above)
Comment (kohlmu-pivotal) [reply geode-core/src/main/java/org/apache/geode/distributed/internal/streaming/StreamingOperation.java]: done
